Re-re-re-re-apply "C++-accessible Placements via pybind11 (#163030)"

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166132

Was reverted (again!) due to a merge conflict that crept in sometime during the "export to github -> land internally -> merge on github" process.

D85096233

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

swolchok committed:
Re-re-re-re-apply "C++-accessible Placements via pybind11 (#163030)"

Was reverted (again!) due to a merge conflict that crept in sometime during the "export to github -> land internally -> merge on github" process.

Differential Revision: [D85096233](https://our.internmc.facebook.com/intern/diff/D85096233/)

[ghstack-poisoned]

ezyang reviewed (approved):
Review automatically exported from Phabricator review in Meta.

swolchok committed:
rebase on "Re-re-re-re-apply "C++-accessible Placements via pybind11 (#163030)""


Was reverted (again!) due to a merge conflict that crept in sometime during the "export to github -> land internally -> merge on github" process.

Differential Revision: [D85096233](https://our.internmc.facebook.com/intern/diff/D85096233/)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
